lib/Dialect/FIRRTL/Transforms/CheckWidths.cpp

```diff
@@ -0,0 +1,106 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the CheckWidths pass.
+//
+// Modeled after:
+// https://github.com/chipsalliance/firrtl/blob/master/src/main/scala/firrtl/passes/CheckWidths.scala
+//
+//===----------------------------------------------------------------------===//
+
+#include "./PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+/// A simple pass that emits errors for any occurrences of `uint`, `sint`, or
+/// `analog` without a known width.
+class CheckWidthsPass : public CheckWidthsBase<CheckWidthsPass> {
+  void runOnOperation() override;
+  void checkType(Type type, Type reportType, Location loc,
+                 llvm::function_ref<void(InFlightDiagnostic &)> addContext);
+
+  InFlightDiagnostic emitError(Location loc, const Twine &message) {
+    signalPassFailure();
+    return mlir::emitError(loc, message);
+  }
+};
+} // namespace
+
+void CheckWidthsPass::runOnOperation() {
+  getOperation()->walk([&](Operation *op) {
+    TypeSwitch<Operation *>(op)
+        .Case<FModuleOp>([&](auto op) {
+          // Check the port types. Unfortunately we don't have an exact location
+          // of the port name, so we just point at the module with the port name
+          // mentioned in the error.
+          for (auto &port : op.getPorts()) {
+            checkType(port.type, port.type, op.getLoc(),
+                      [&](InFlightDiagnostic &error) {
+                        error.attachNote()
+                            << "in port `" << port.getName() << "` of module `"
+                            << op.getName() << "`";
+                      });
+          }
+        })
+        .Default([&](auto *op) {
+          // Check each result of the operation.
+          for (auto type : op->getResultTypes())
+            checkType(type, type, op->getLoc(), [&](InFlightDiagnostic &error) {
+              error.attachNote() << "in result of `" << op->getName() << "`";
+            });
+        });
+  });
+}
+
+/// Check a type and its children for any occurrences of uninferred widths.
+/// Specifically look for `uint`, `sint`, and `analog` types. The argument
+/// `type` is the one being checked, `reportType` is the one included in any
+/// errors, `loc` is the location to report errors to, and `addContext` gives
+/// the caller the chance to add additional context information to the error.
+void CheckWidthsPass::checkType(
+    Type type, Type reportType, Location loc,
+    llvm::function_ref<void(InFlightDiagnostic &)> addContext) {
+  TypeSwitch<Type>(type)
```

Comment:
I'd recommend adding a `DenseSet<Type>` to CheckWidthsPass to keep track of types that have already been checked for being ok.  This avoids problems where deep bundle types could burn time unnecessarily.